### ISSUE TITLE
Multiclass Confmat calculation throws different device exception

### ISSUE THREAD

**Issue by onurtore** [Lightning-AI/torchmetrics#1865]
## 🐛 Bug

When you try to calculate the confmat with GPU tensors for multiclass it throws the different device exception. Problems arises due to below operation:

`self.confmat += conftmat
`
Where confmat defined in the GPU, and self.conftmat defined in CPU. Moving self.conftmat to GPU using self.conftmat.cuda() solves this issue.

### To Reproduce

Try to calculate multiclass conftmat using GPU tensors. Pred dims (N,C,H,W), Label dim (N,H,W)

<details>
  <summary>Code sample</summary>

</details>

### Environment

- TorchMetrics version (and how you installed TM, e.g. `conda`, `pip`, build from source):
- Python & PyTorch Version (e.g., 1.0):
- Any other relevant information such as OS (e.g., Linux):

**Comment by SkafteNicki** [comment]
Hi @onurtore, thanks for raising this issue.
I assume the problem here is that you are not moving the metric to GPU before you try to use it?
Metrics are just simple `nn.Module` thus they need to be moved to the device where calculations are going to happen. You can read more about this here:
https://torchmetrics.readthedocs.io/en/stable/pages/overview.html#metrics-and-devices

And here is a simple example:
```python
import torch
from torchmetrics.classification import MulticlassConfusionMatrix

num_classes = 10

preds = torch.randn(100, num_classes, 28, 28).softmax(dim=1)
target = torch.randint(num_classes, (100, 28, 28))

metric = MulticlassConfusionMatrix(num_classes=num_classes)
assert metric.confmat.device == torch.device("cpu")

# move to GPU
metric = metric.cuda()
assert metric.confmat.device == torch.device("cuda:0")

result = metric(preds.cuda(), target.cuda())
assert result.device == torch.device("cuda:0")
```
:]

**Comment by SkafteNicki** [comment]
Sorry if I did not understand the question. The short answer is that you should not be able to do it for F1 and Accuracy either, that is an error. Locally it also works as expected for me, namely an error telling me that my accuracy metric has not been moved to the correct device.
Running:
```python
import torch
from torchmetrics.classification import MulticlassAccuracy

num_classes = 10

preds = torch.randn(100, num_classes, 28, 28).softmax(dim=1)
target = torch.randint(num_classes, (100, 28, 28))

metric = MulticlassAccuracy(num_classes=num_classes)
result = metric(preds.cuda(), target.cuda())
```
gives me
![image](https://github.com/Lightning-AI/torchmetrics/assets/24896311/aa5e2e7a-7809-4cc4-bfab-f79a328faf30)
